chalicelib/utils.py

```diff
@@ -44,13 +47,29 @@ def __init__(self, data, url, workspace, namespace, auth=None):
         }
 
     def workspace_exists(self):
-        """Returns true if FireCloud workspace in the specified namespace
-        exists, otherwise false.
-        """
+        """Returns requests response object. Status code is 200 if FireCloud 
+        workspace in the specified workspace exists, or 404 if it doesn't."""
         url = self.workspace_url
         headers = dict(Accept='application/json',
                        Authorization=self.auth)
-        return requests.get(url=url, headers=headers)
+        # Start a session with 3 GET retries.
+        s = Session()
+        # The reason for the 401 in the status_forcelist is because we
+        # have seen it occurring in FC requests.
+        try:
+            s.mount('https://', HTTPAdapter(
+                max_retries=Retry(total=3,
+                                  backoff_factor=1,
+                                  redirect=False,
+                                  status_forcelist=[401, 500, 502, 503, 504])
+            ))
+            return s.get(url=url, headers=headers)
+        except Exception as x:
+            # after three retrys:
+            return Response(
+                body={'Request failed:': x.__class__.__name__},
```

Comment:
What is this value typically?

It would be good to have a more meaningful message.

More importantly, I'm pretty sure this will return a 200 status code, which you definitely don't want. If there is an error you need to return an error to the client

Reply:
True, it returned a 200. I changed it such that it returns the RetryError message.